reactive_graph/src/traits.rs

```diff
@@ -639,3 +639,199 @@ pub fn panic_getting_disposed_signal(
+    fn with_value<U>(&self, fun: impl FnOnce(&Self::Value) -> U) -> U {
+        self.try_with_value(fun)
+            .unwrap_or_else(unwrap_signal!(self))
+    }
+}
+
+impl<T> WithValue for T
+where
+    T: DefinedAt + ReadValue,
+{
+    type Value = <<Self as ReadValue>::Value as Deref>::Target;
+
+    fn try_with_value<U>(
+        &self,
+        fun: impl FnOnce(&Self::Value) -> U,
+    ) -> Option<U> {
+        self.try_read_value().map(|value| fun(&value))
+    }
+}
+
+/// A variation of the [`Get`] trait that provides a signposted "always-non-reactive" API.
+/// E.g. for [`StoredValue`](`crate::owner::StoredValue`).
+pub trait GetValue: DefinedAt {
+    /// The type of the value contained in the signal.
+    type Value: Clone;
+
+    /// Clones and returns the value of the signal, non-reactively,
+    /// or `None` if the signal has already been disposed.
+    #[track_caller]
+    fn try_get_value(&self) -> Option<Self::Value>;
+
+    /// Clones and returns the value of the signal, non-reactively.
+    ///
+    /// # Panics
+    /// Panics if you try to access a signal that has been disposed.
+    #[track_caller]
+    fn get_value(&self) -> Self::Value {
+        self.try_get_value().unwrap_or_else(unwrap_signal!(self))
+    }
+}
+
+impl<T> GetValue for T
+where
+    T: WithValue,
+    T::Value: Clone,
+{
+    type Value = <Self as WithValue>::Value;
+
+    fn try_get_value(&self) -> Option<Self::Value> {
+        self.try_with_value(Self::Value::clone)
+    }
+}
+
+/// A variation of the [`Write`] trait that provides a signposted "always-non-reactive" API.
+/// E.g. for [`StoredValue`](`crate::owner::StoredValue`).
+pub trait WriteValue: Sized + DefinedAt {
+    /// The type of the signal's value.
+    type Value: Sized + 'static;
+
+    /// Returns a non-reactive write guard, or `None` if the signal has already been disposed.
+    #[track_caller]
+    fn try_write_value(&self) -> Option<UntrackedWriteGuard<Self::Value>>;
+
+    /// Returns a non-reactive write guard.
+    ///
+    /// # Panics
+    /// Panics if you try to access a signal that has been disposed.
+    #[track_caller]
+    fn write_value(&self) -> UntrackedWriteGuard<Self::Value> {
+        self.try_write_value().unwrap_or_else(unwrap_signal!(self))
+    }
+}
+
+/// A variation of the [`Update`] trait that provides a signposted "always-non-reactive" API.
+/// E.g. for [`StoredValue`](`crate::owner::StoredValue`).
+pub trait UpdateValue: DefinedAt {
+    /// The type of the value contained in the signal.
+    type Value;
+
+    /// Updates the value by applying a function, non-reactively, returning the value returned by that function,
+    /// or `None` if the signal has already been disposed.
+    #[track_caller]
+    fn try_update_value<U>(
+        &self,
+        fun: impl FnOnce(&mut Self::Value) -> U,
+    ) -> Option<U>;
+
+    /// Updates the value by applying a function, non-reactively, returning the value returned by that function.
+    ///
+    /// # Panics
```

Comment:
Current behavior in both 0.6 and 0.7 is that `update_value` 1) doesn't return `U` and therefore 2) is guaranteed never to panic -- if the value has already been disposed it simply does nothing. I think we should stick to this pattern.

Reply:
Yep fair enough, I think I'd copied from StoredValue so the API had changed, but for the set one I was thinking about it as if each disposed error is a downstream user error hence they should see it straight away. But yeah makes sense to avoid an error when possible. Updated